File: dotnet/DesignPatterns/StructuralDesignPatterns/FacadeDesignPattern/FacadeDesignPattern.Example1/HomeTheaterFacade.cs
using FacadeDesignPattern.Example1.Models;

namespace FacadeDesignPattern.Example1;

public class HomeTheaterFacade
{
	private readonly Projector _projector;
	private readonly Amplifier _amplifier;
	private readonly DvdPlayer _dvdPlayer;
	private readonly Lights _lights;

	public HomeTheaterFacade(Projector projector, Amplifier amplifier, DvdPlayer dvdPlayer, Lights lights)
	{
		_projector = projector;
		_amplifier = amplifier;
		_dvdPlayer = dvdPlayer;
		_lights = lights;
	}

	public void WatchMovie(string movie)
	{
		Console.WriteLine("\n--- Starting Movie Mode ---");
		_lights.Dim();
		_projector.On();
		_projector.SetWideScreenMode();
		_amplifier.On();
		_amplifier.SetVolume(10);
		_dvdPlayer.On();
		_dvdPlayer.Play(movie);
	}

	public void EndMovie()
	{
		Console.WriteLine("\n--- Shutting Down Movie Mode ---");
		_dvdPlayer.Stop();
		_dvdPlayer.Off();
		_amplifier.Off();
		_projector.Off();
		_lights.On();
	}
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/FacadeDesignPattern/FacadeDesignPattern.Example1/src/HomeTheaterFacade.ts
import { Projector } from './models/Projector';
import { Amplifier } from './models/Amplifier';
import { DvdPlayer } from './models/DvdPlayer';
import { Lights } from './models/Lights';

export class HomeTheaterFacade {
  constructor(
    private readonly projector: Projector,
    private readonly amplifier: Amplifier,
    private readonly dvdPlayer: DvdPlayer,
    private readonly lights: Lights,
  ) {}

  watchMovie(movie: string): void {
    console.log('\n--- Starting Movie Mode ---');
    this.lights.dim();
    this.projector.on();
    this.projector.setWideScreenMode();
    this.amplifier.on();
    this.amplifier.setVolume(10);
    this.dvdPlayer.on();
    this.dvdPlayer.play(movie);
  }

  endMovie(): void {
    console.log('\n--- Shutting Down Movie Mode ---');
    this.dvdPlayer.stop();
    this.dvdPlayer.off();
    this.amplifier.off();
    this.projector.off();
    this.lights.on();
  }
}
